### ISSUE TITLE
Hadamard (BitNet v2) training diverges after ~2500 steps

### ISSUE THREAD

**Issue by Lev-Stambler** [DeepOpt-com/WrinkleFree-Monorepo#51]
## Problem

Training with HBitLinear (Hadamard transform + ternary quantization) consistently diverges after ~2000-2500 steps, despite initial loss improvement.

## Observed Behavior

| Phase | Steps | Train Loss | Val Loss | Notes |
|-------|-------|------------|----------|-------|
| Lambda warmup | 0-200 | 7.0 → 6.5 | - | Improving ✓ |
| Post-warmup | 200-2000 | 6.5 → 5.3 | 5.8 | Still improving ✓ |
| Instability | 2000-2500 | 5.3 → 6.5 | 6.4 | Creeping up ⚠️ |
| Divergence | 2500+ | 6.5 → 12+ | 12+ | Explodes ❌ |

## Configuration

- **Model**: Qwen3-0.6B (hidden=1024, intermediate=3072)
- **HBitLinear layers**: o_proj, down_proj only
- **LR**: 5e-5 (AdamW)
- **Lambda warmup**: 200 steps (also tried 500)
- **torch_compile**: enabled
- **DLM**: disabled (CE only)

## Key Dimensions

```
hidden_size: 1024      # Power of 2 ✓
intermediate_size: 3072  # NOT power of 2 → pads to 4096
```

The `down_proj` layer has `in_features=3072` which gets padded to 4096 for Hadamard transform.

## Fixes Applied (torch.compile hanging)

1. **HBitLinear buffer fix**: Changed from global `_hadamard_cache` dict to registered buffer in `__init__`
   - File: `packages/architecture/src/wf_arch/layers/hbitlinear.py`
   - This fixed torch.compile hanging indefinitely

2. **Lambda warmup decorator**: Added `@torch.compiler.disable` to `get_current_lambda()`
   - File: `packages/architecture/src/wf_arch/quantization/lambda_warmup.py`

## Hypotheses for Divergence

1. **LR too high**: 5e-5 may be too aggressive for Hadamard + quantization
2. **Non-power-of-2 padding issue**: Mathematical equivalence may break when slicing back from padded dimensions
3. **Quantization + Hadamard interaction**: The combination of ternary weights + Hadamard-transformed activations may be unstable
4. **Missing gradient clipping**: No ZClip or gradient norm clipping configured

## Proposed Experiments

- [ ] Try lower LR (1e-5, 2e-5)
- [ ] Try longer lambda warmup (1000+ steps)  
- [ ] Add gradient clipping
- [ ] Test with power-of-2 intermediate size model
- [ ] Compare against non-Hadamard BitLinear baseline
- [ ] Verify mathematical equivalence holds during training (not just init)

## Related Files

- `packages/architecture/src/wf_arch/layers/hbitlinear.py`
- `packages/architecture/src/wf_arch/layers/hadamard.py`
- `packages/deployer/skypilot/train_hadamard_ce_only.yaml`

## WandB Runs

- https://wandb.ai/umd-leans-well/wrinklefree_v2/runs/hadamard_qwen3_0.6b_ce_only-dc729208 (diverged)
- https://wandb.ai/umd-leans-well/wrinklefree_v2/runs/hadamard_qwen3_0.6b_ce_only-fce5349a (killed early)
